opentelemetry-etw-metrics/src/exporter/mod.rs

```diff
@@ -39,265 +41,146 @@ impl Debug for MetricsExporter {
     }
 }
 
+fn emit_export_metric_service_request(
+    export_metric_service_request: &ExportMetricsServiceRequest,
+    encoding_buffer: &mut Vec<u8>,
+) -> MetricResult<()> {
+    if (export_metric_service_request.encoded_len()) > etw::MAX_EVENT_SIZE {
+        otel_warn!(name: "MetricExportFailedDueToMaxSizeLimit", size = export_metric_service_request.encoded_len(), max_size = etw::MAX_EVENT_SIZE);
+    } else {
+        encoding_buffer.resize_with(
+            export_metric_service_request.encoded_len(),
+            Default::default,
+        );
+
+        export_metric_service_request
+            .encode(encoding_buffer)
+            .map_err(|err| MetricError::Other(err.to_string()))?;
+
+        let result = etw::write(encoding_buffer);
+        // TODO: Better logging/internal metrics needed here for non-failure
+        // case Uncomment the line below to see the exported bytes until a
+        // better logging solution is implemented
+        // println!("Exported {} bytes to ETW", byte_array.len());
+        if result != 0 {
+            otel_warn!(name: "MetricExportFailed", error_code = result);
+        }
+    }
+
+    Ok(())
+}
+
 #[async_trait]
 impl PushMetricExporter for MetricsExporter {
     async fn export(&self, metrics: &mut ResourceMetrics) -> MetricResult<()> {
+        let schema_url: String = metrics
+            .resource
+            .schema_url()
+            .map(Into::into)
+            .unwrap_or_default();
+
+        let mut encoding_buffer = Vec::new();
```

Comment:
This can be optimized by using a pre-allocated thread-local vector of MAX_EVENT_SIZE size. Probably for future optimization.

Reply:
Related to this https://github.com/open-telemetry/opentelemetry-rust-contrib/pull/134#discussion_r1914018851.

I think it might be better to use a `Mutex` protected pre-allocated `vec` instead of a thread local `vec`. If the users are setting up this exporter with a runtime such as Tokio, then having a thread local `vec` would mean that potentially all worker threads would end up allocating a `vec` of size MAX_EVENT_SIZE. If we use a `Mutex` protect thread instead, we will only allocate this `vec` once and there would be no contention for the `Mutex` as we never call export concurrently.

Reply:
Oh missed that comment. Btw, export can be called concurrently as this is the re-entrant processor right?

Reply:
No with metrics we always "batch" (aggregate the data) and export it. If an export is already in progress, the subsequent export would wait for the previous export to complete.

Reply:
Oh yes, was mixing logs and metrics :)